site/_posts/2018-02-12-discovering-a-ruby-memory-leak.html.md

```diff
@@ -0,0 +1,441 @@
+to take a look at the core dumps as hex and determine where patterns of leaked
+memory were.  We weren't that smart to be able to understand that ourselves or
+make use of it properly in our case, but we did end up using the
+`rb_eval_string("")` portions to our advantage though, and that ended up being
+super useful.
+
+At the time, we were under the impression that `MiqServer#sync_workers` was the
+culprit, so we ended up attaching to an existing 'warmed up' server process
+that had been running for a bit so most of the code it would load would already
+be autoloaded.  With that, we tried running a high number of invocations of
+that method in a row, and watched the `/proc/<PID>/smaps` to see if the
+`[heap]` section grew.  It looked something like this\*\*\*:
+
+```console
+$ gdb -p [PID_OF_THE_MIQ_SERVER_PROCESS]
+(gdb) p rb_eval_string("GC.start")
+(gdb) p rb_eval_string("500.times { MiqServer.my_server.sync_workers; GC.start }")
+```
+
+Nothing.  Not even a budge.  Well shoot... there goes that theory.  
+
+At this point, we looked back to the servers we had running that were modified
+with the  known "non-leaking" portions commented out, and checked if there was
+anything else that we might have been missing.  We noticed that in addition to
+`MiqServer#sync_workers` still being uncommented, a call to
+`MiqWorker.status_update_all` that was directly preceding it was also still
+being executed.
+
+Now, on it's own, this method looks seemed pretty tame. Effectively it:
+
+1. Uses basic ActiveRecord to fetch all the workers under the current server 
+2. Makes a call to our own code to fetch usage data for the given worker pids
+3. Saves the data to the DB records for internal reporting/monitoring
+
+The first and last bits were ruled out, and we were pretty sure we already
+tested the middle bit via some scripts outside of the context of ManageIQ, but
+for verbosity sake, using the same `gdb` process, we decided to update the
+`500.times {}` block to include that call:
+
+```ruby
+MiqServer.my_server.sync_workers; MiqWorker.status_update_all; GC.start 
+```
+
+And... it leaked... !?
+
+
+
+Step 5:  Wait... wut!? (aka:  replicated it without our app)
+------------------------------------------------------------
+
+So, at this point, we started to look at the call to
+`MiqProcess.processInfo(worker_pid)` in `MiqWorker.status_update_all` since the
+other areas are mostly ActiveRecord.  Besides, it's easier to blame our own
+libraries. ;)
+
+Now, before you say "good move dummies, that was your code... why didn't try
+that in the first place", let me explain.  We had tried something similar in a
+test script with no leak, but ended up trying it again.  The first draft of the
+script looked basically like this:
+
+```ruby
+# Code needed to get MiqProcess.processInfo loaded and working
+require 'manageiq-gems-pending'
+require 'util/miq-process'
+require 'active_support/all'
```

Comment:
Remove code that isn't pertinent. The reader won't even notice. 99.9999% of people reading this aren't going to literally paste this in a script and be mad when it doesn't work as-is.

Reply:
Going to disagree here since I don't think this makes the script any harder to read for anyone that is a rubist and gotten this far.  Also, it also is intended to re-enforce the point made further down in which we literally added a `require 'config/environment'` to this that changed the script and that caused it to start leaking, and having the context of this originally being attempted outside the app is important.

Also, requires are pretty much something that is glossed over for anyone speeding through this, and I don't think it really would significantly help with the culling effort.